### PR TITLE
GitCommitBear.py: Issue reference regex matching

### PR DESCRIPTION
**Pull request by nkprince007**

### CODE REVIEW

File: bears/vcs/git/GitCommitBear.py [hunk 1]
```diff
@@ -203,3 +243,82 @@ def check_body(self, body,
...
+            result_message = ('Body of HEAD commit does not contain any {} '
+                              'reference.')
+
+        if body_close_issue_full_url:
+            result_info = 'full issue'
+            compiled_issue_ref_regex = re.compile(
+                r'https?://{}\S+/issues/\w+'.format(re.escape(host)))
+        else:
+            result_info = 'issue'
+            compiled_issue_ref_regex = re.compile(r'#\w+')
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: you should assign first a string, and compile the regex outside the if, avoids code duplication :+1:

Comment (nkprince007) [inline bears/vcs/git/GitCommitBear.py]: Ok. 👍

File: bears/vcs/git/GitCommitBear.py [hunk 2]
```diff
@@ -203,3 +243,82 @@ def check_body(self, body,
...
+                self.SUPPORTED_HOST_KEYWORD_REGEX[host],
+                concat_regex))
+
+        matches = compiled_joint_regex.findall(body)
+
+        if len(matches) == 0:
+            yield Result(self, result_message.format(result_info))
+            return
+
+        compiled_issue_num_regex = re.compile(r'(?:#|\D+/)[1-9][0-9]*')
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: I think you could use the regex without the `(?:#|\D+/)` part. You could add a capturing group inside `compiled_issue_ref_regex` getting the right part of the reference that represent the issue number. Inside the check-loop below, you extract the group and just try a fullmatch with `[1-9][0-9]*`.

what do you think?

File: bears/vcs/git/GitCommitBear.py [hunk 3]
```diff
@@ -203,3 +243,82 @@ def check_body(self, body,
...
+                r'https?://{}\S+/issues/\w+'.format(re.escape(host)))
+        else:
+            result_info = 'issue'
+            compiled_issue_ref_regex = re.compile(r'#\w+')
+
+        concat_regex = '|'.join(kw for kw in self.CONCATENATION_KEYWORDS)
+        compiled_joint_regex = re.compile(
+            r'(?:{0})\s+'
+            r'((?:\S(?!{1}))*\S(?:\s*(?:{1})\s*(?!{0})(?:\S(?!{1}))*\S)*)'
+            r''.format(
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: like mentioned please describe the regex ;)

File: bears/vcs/git/GitCommitBear.py [hunk 4]
```diff
@@ -3,6 +3,8 @@
 import shutil
 import os
 
+from urllib.parse import urlparse
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: it's a system import, so no newline between `import os` and this one :)

File: bears/vcs/git/GitCommitBear.py [hunk 5]
```diff
@@ -203,3 +243,94 @@ def check_body(self, body,
...
+            return
+
+        compiled_issue_ref_regex = re.compile(issue_ref_regex)
+        compiled_issue_no_regex = re.compile(r'[1-9][0-9]*')
+        compiled_concat_regex = re.compile(
+            r'\s*(?:{})\s*'.format(concat_regex))
+
+        for match in matches:
+            for issue in re.split(compiled_concat_regex, match):
+                reference = compiled_issue_ref_regex.match(issue)
```
Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: it should fullmatch^^

Comment (nkprince007) [inline bears/vcs/git/GitCommitBear.py]: Oh, yeah. :O

Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: example when it won't work:
```
Fixes #123-4
```
(as `\w` doesn't match `-`. We should maybe use `\S`^^)

Comment (Makman2) [inline bears/vcs/git/GitCommitBear.py]: and here you get another test case :P

**Comment by Makman2** [comment]
ack 43b70c3, 47a0932

**Comment by Makman2** [comment]
ack e433f3e98f49940956fe2fe98c4d482888d8c07f 6a33fbaa4b92bf6891a1adaa29881f0549518691 ae565b7eeebf7b0c188f0a72a915f2cd1ed0dd28

**Comment by Makman2** [comment]
@rultor merge

**Pull request merged**